common/page.go

```diff
@@ -70,13 +78,19 @@ type Page struct {
 
 	backgroundPage bool
 
+	eventCh         chan Event
+	eventHandlers   map[string][]consoleEventHandlerFunc
+	eventHandlersMu sync.RWMutex
+
 	mainFrameSession *FrameSession
 	frameSessions    map[cdp.FrameID]*FrameSession
 	frameSessionsMu  sync.RWMutex
 	workers          map[target.SessionID]*Worker
 	routes           []api.Route
 	vu               k6modules.VU
 
+	tq *taskqueue.TaskQueue
```

Comment:
A couple of comments about the usage of the `TaskQueue`:

1. The `TaskQueue` must be closed in order to let the event loop finish and therefore the iteration to finish. Considering that now the browser is abstracted from the JS API, this will only happen if the actual chrome instance exits for some reason, or if page is explicitly closed in the test. Which means that users have to call `page.close()` to let the iteration finish. I don't see other event that would allow us to close the task queue safely (see point 2 also). One side effect of this is that requiring the explicit `page.close()` we can guarantee also the Web Vitals "force" measurements.
The only improvement that I can think of about this is to only initialize the `TasQueue` when/if `page.on` is called. Therefore the user would only be required to close the page if using this method. But this might make it more confusing overall than always requiring to close the page...
2. We had a brief discussion on making the `TaskQueue` a global entity, possibly being handled at the mapping layer/registry. In practice I think this is not an option, as there is not an event that we can hook to in order to close the queue. The only event that would make sense is the `IterEnd` event, but this will not be fired from k6 until the iteration ends, and the iteration will not end until the event loop ends, which requires the `TaskQueue` to be closed.
All in all I think it makes sense to use it locally in the `page` implementation, especially now that we only have one use case for it in the whole codebase, and also considering how easy it us to use it.

Reply:
I think this might be okay for now. 👍

Perhaps, if we consider even a single use case, there might be an opportunity later on to shift this task queue and event-loop-related logic into the mapping layer. For example, we [moved the promises (event-loop-related logic) to the mapping layer](https://github.com/grafana/xk6-browser/issues/683), which helped in simplifying our tests and code. We're also [looking into the idea of moving the Goja-related mechanics](https://github.com/grafana/xk6-browser/issues/271) away from the business logic to the mapping layer, aiming for improved maintainability. I hope we're continuing in the direction of progress.

Reply:
I think you've done a great job on incorporating the taskqueue into k6 browser 👏 Looks super useful in general.

My main worry is that we're now relying on the user to use `page.close`, which is probably an ok requirement, but it feels like a poor DX if k6 hangs and times out (i'm assuming it's a timeout) if the user forgets to work with `page.close`. It would be great if we could handle the lifecycle of the eventqueue ourselves, but considering the taskqueue must be closed before the iterEnd can be fired by k6, I can't see how this can be done 🤔 

Could we have a taskqueue for the whole test run and wait for the `Exit` event (although doesn't feel like the correct fit)? I'm presuming the `Exit` event will fire regardless of whether the taskqueue is closed 🤷

Reply:
We can reevaluate having the `TaskQueue` as a "global component", but again, I don't see how could we handle its closing in a better way.

> Could we have a taskqueue for the whole test run and wait for the Exit event

I believe here the problem is that the `Exit` event will not be fired until all iterations have ended, and the iteration won't end if the task queue is not closed. So we have the same issue as waiting for the `IterEnd`.

The only alternative that I could think of, to not have to use the `TaskQueue`, was to limit the `page.on` method to receive just a single `msg` per call. Or allow the user to define how many messages does he/she want to receive. In these cases I believe we could get rid of the task queue. However, I think these alternatives are worse than requiring the `page.close()` call.

Reply:
> We had a brief discussion on making the TaskQueue a global entity, possibly being handled at the mapping layer/registry. In practice I think this is not an option, as there is not an event that we can hook to in order to close the queue.
> I don't see how could we handle its closing in a better way.

What about adding the task queue to the registry, then in the mapping layer, creating a new task queue on a new page(s) call(s), queue `page.On` handlers there, and closing the task queue when `page.close` is called? We might want to make it page specific, though. And, this isn't lifting the `page.Close` problem, though. For that, it seems we need k6-core's help, as I cannot see another way to do this either :(

Reply:
> What about adding the task queue to the registry (as a field in module vu), then in the mapping layer, creating a new task queue on a new page(s) call(s), queue page.On handlers there, and closing the task queue when page.close is called?

I see pros and cons of this. From one side we would remove the usage of the `TaskQueue` (wich is an element mainly related with Goja) from our core implementation, which is something we are trying to do. But at the same time I think it would increase the complexity of the codebase. Especially considering how easy it has been to use the `TaskQueue` in our `page` implementation, as it basically adds "3" lines of code. For example, I think a similar situation is the one that we faced with the mapping for the `page.on` events. In the end this was fixed like [this](https://github.com/grafana/xk6-browser/blob/1de4e41f70cb5a3c9dfabd71f3640631a9696e2e/browser/mapping.go#L510-L517), which I think is quite complex to understand, but in this case there was no better solution.

For the options that we have considered, I would keep the usage of `TaskQueue` at the `page` implementation by now, unless we can find a global implementation that will cover more use cases (as now we have only one), or another implementation that would allow us to not require an explicit `page.close` call.

Reply:
Yes, Goja is complex, and event-loop is even more; that's why I prefer to keep it away from the actual code. Not only does the code, but also complicates the tests a lot, which we see [here](https://github.com/grafana/xk6-browser/pull/1006#discussion_r1305498962) and [here](https://github.com/grafana/xk6-browser/pull/1006#discussion_r1305513979). It takes away the expressiveness of tests and adds a lot of boilerplate. As I said, I'm fine with this for now until we find a better solution.

Reply:
I'm also happy with the current solution, thanks @ka3de!